src/harness/reference_models/pre_iap_filtering/fss_purge.py

```diff
@@ -0,0 +1,301 @@
+  fss_entity = interference.FssProtectionPoint(latitude=fss_latitude,
+                                               longitude=fss_longitude,
+                                               height_agl=fss_height_agl,
+                                               max_gain_dbi=fss_max_gain_dbi,
+                                               pointing_azimuth=fss_azimuth,
+                                               pointing_elevation=fss_pointing_elevation)
+  return fss_entity
+
+def performPurge(cbsds, fss_entity):
+  """ Implementation of algorithm described in R2-SGN-29.
+
+  The list of grants to be purged based on the interference caused by their
+  out of band emission are identified and returned.
+  Args:
+    cbsds: List of CbsdData objects that are in the neighborhood to an FSS.
+    fss_entity: The FssProtectionPoint tuple of the FSS for which the purge is to be performed.
+  Returns:
+    List of multiple key-value pairs with keys "grantData", "cbsd" and values as grants to be
+      purged and corresponding cbsd data object respectively.
+  """
+  # List of grants to be purged for the fss_entity.
+  final_purge_list = []
+
+  # List of dictionaries containing the highest grant of the CBSD and a reference to the CBSD.
+  grants_for_oobe_calculation = getHighestGrantOfCbsds(cbsds)
+
+  # Calculate the threshold value in dBm for the OOBE interference
+  oobe_threshold_value = -129 + 10 * (np.log10(REF_BW/interference.ONE_MHZ) - FSS_OOBE_MARGIN)
+
+  while True:
+    # Calculate the OOBE interference value for each grant.
+    calculateOobeInterference(grants_for_oobe_calculation, fss_entity)
+
+    # Sort the grants based on the calculated interference value.
+    sorted_grants = sorted(grants_for_oobe_calculation, key=lambda grant: grant['oobe_interference'])
+
+    # Find the largest index such that aggregation of the interferences do not cross
+    # the OOBE threshold.
+    oobe_interferences = [interference.dbToLinear(grant['oobe_interference']) for grant in sorted_grants]
+    cumulated_interference = np.cumsum(oobe_interferences)
+    index = np.searchsorted(cumulated_interference, interference.dbToLinear(oobe_threshold_value))
+    grants_to_purge = sorted_grants[index:]
+    grants_for_oobe_calculation = sorted_grants[:index]
+    final_purge_list.extend(grants_to_purge)
+
+    # Add the identified grants to be purged from this iteration to the final list of grants to be purged
+    if grants_to_purge:
+      for purged_grant in grants_to_purge:
+        remaining_grants = []
+        # Identify other grants of the CBSD that have the same emission mask as the purged
+        # grant and add them to the final purge list. If the CBSD has any other grant with
+        # lower emission mask then add it to the remaining_grants list.
+        for grant in purged_grant['cbsd']['grants']:
+          # Check that we are not comapring the same grant as the pruged grant
+          if grant != purged_grant['grantData']:
```

Comment:
This is creating a special case without a good reason.

Reply:
In this   function we wanted to identify other grants of cbsd with the same emission   mask, extra check is introduced to avoid grant to compare with same(self)   grant

Reply:
My point is there is no need to do this check. Removing the bit at the start (now lines 110-111) that adds all the directly referenced grants to ```final_purge_list``` and then removing this check will make the code easier to read while doing exactly the same thing.